rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/model/PathConfigDiagnostics.java

```diff
@@ -0,0 +1,141 @@
+        this.client = client;
+        this.cm = cm;
+    }
+
+    public void publishDiagnostics(ISourceLocation project, IList messages) {
+        Set<ISourceLocation> filesToRepublish = new HashSet<>();
+
+        // Gather messages per file
+        Map<ISourceLocation, List<Diagnostic>> diagnostics = Diagnostics.translateMessages(messages, cm);
+
+        // Do the actual manipulation of the internal datastructures in a synchronized block
+        // so their coherence is guaranteed.
+        synchronized(this) {
+            // Publish diagnostics, but only for files for which diagnostics have changed
+            for (var entry : diagnostics.entrySet()) {
+                ISourceLocation file = entry.getKey().top();
+                List<Diagnostic> newDiagnostics = entry.getValue();
+
+                Map<ISourceLocation, List<Diagnostic>> diagsPerProject = projectDiagsPerFile.computeIfAbsent(file, (f) -> new LinkedHashMap<>());
+                @Nullable List<Diagnostic> publishedDiags = diagsPerProject.get(project);
+
+                // Only publish diagnostics for a file if the diagnostics for that file have changed in the current project
+                if (!newDiagnostics.equals(publishedDiags)) {
+                    // The diagnostics for this file related to the current project have changed.
+                    // We need to re-publish all diagnostics for this file (including the ones from other projects)
+                    diagsPerProject.put(project, newDiagnostics);
+                    filesToRepublish.add(file);
+                }
+            }
+
+            // Now we have published the diagnostics of all files that occur in the list of new messages
+            // Time to remove diagnostics from files that our project previously published diagnostics for but no longer have diagnostics
+            Set<ISourceLocation> newFiles = diagnostics.keySet().stream().map(file -> file.top()).collect(Collectors.toSet());
+            Set<ISourceLocation> filesWithoutDiags = new HashSet<>(filesWithDiagsPerProject.getOrDefault(project, Collections.emptySet()));
+            filesWithoutDiags.removeAll(newFiles);
+
+            for (ISourceLocation file : filesWithoutDiags) {
+                // File no longer has diagnostics associated with it in our project
+                projectDiagsPerFile.get(file).remove(project);
+                filesToRepublish.add(file);
+            }
+            filesWithDiagsPerProject.put(project, newFiles);
+        }
+
+        publishFileDiagnostics(filesToRepublish);
+    }
+
+    public void clearDiagnostics(ISourceLocation project) {
+        Set<ISourceLocation> affectedFiles;
+
+        synchronized (this) {
+            affectedFiles = filesWithDiagsPerProject.remove(project);
+            if (affectedFiles == null) {
+                return;
+            }
+
+            for (ISourceLocation file : affectedFiles) {
+                projectDiagsPerFile.get(file).remove(project);
```

Comment:
if we start cleaning up the `projectDiagsPerFile` map, we have to become more defensive as `get` might return a null.